server/lib/routes/post-csp.js

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Route to report CSP Violations to metrics
+ */
+
+var DEFAULT_RATE_LIMIT = 90;
+
+module.exports = function (options) {
+  options = options || {};
+
+  /**
+   * Rate limit 'rateLimitPercentage' % of messages.
+   * @param rateLimitPercentage
+   * @returns {boolean}
+   */
+  function isRateLimited(rateLimitPercentage) {
+    var RATE_LIMIT = rateLimitPercentage;
+    // rate limit can be 0
+    if (typeof rateLimitPercentage === 'undefined') {
+      RATE_LIMIT = DEFAULT_RATE_LIMIT;
+    }
+    var min = 0;
+    var max = 100;
+
+    var rand = min + Math.floor(Math.random() * (max - min + 1));
+    return rand <= RATE_LIMIT;
+  }
+
+  return {
+    method: 'post',
+    path: '/_/csp-violation',
+    process: function (req, res) {
+      res.json({result: 'ok'});
+
+      // TODO: This is a temporary measure
+      // Not sure how many CSP errors we will get, for now we rate limit this.
+      // To avoid overflowing Heka logs rate limit the logging
+      if (isRateLimited(options.rateLimitPercentage)) {
+        return false;
+      }
+
+      if (! req.body || ! req.body['csp-report'] || ! req.body['csp-report']['blocked-uri']) {
+        return false;
+      }
+
+      var today = new Date();
+      today.setMinutes(0, 0, 0);
+
+      var entry = {
+        blocked: req.body['csp-report']['blocked-uri'],
```

Comment:
Looking at the [CSP spec](https://w3c.github.io/webappsec-csp/#deprecated-serialize-violation), it may be useful to know the `referrer` of the offending script. I have a suspicion that addons like LastPass will cause CSP violations if they insert scripts into the DOM automatically, and we'll want to know that.